core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.metrics;
+
+import alluxio.Configuration;
+import alluxio.PropertyKey;
+import alluxio.client.file.FileSystemContext;
+import alluxio.heartbeat.HeartbeatExecutor;
+import alluxio.metrics.Metric;
+import alluxio.metrics.MetricsSystem;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Task that carries the client metrics information to master through heartheat. This class manages
+ * its own {@link MetricsMasterClient}.
+ *
+ * If the task fails to heartbeat to the master, it will destroy its old master client and recreate
+ * it before retrying.
+ */
+@ThreadSafe
+public final class ClientMasterSync implements HeartbeatExecutor {
+  private static final Logger LOG = LoggerFactory.getLogger(ClientMasterSync.class);
+
+  /** Client for communicating to metrics master. */
+  private final MetricsMasterClient mMasterClient;
+
+  /** Milliseconds between heartbeats before a timeout. */
+  private final int mHeartbeatTimeoutMs;
+
+  /** Last System.currentTimeMillis() timestamp when a heartbeat successfully completed. */
+  private long mLastSuccessfulHeartbeatMs;
+
+  /**
+   * Constructs a new {@link ClientMasterSync}.
+   *
+   * @param masterClient the master client
+   */
+  public ClientMasterSync(MetricsMasterClient masterClient) {
+    mMasterClient = Preconditions.checkNotNull(masterClient, "masterClient");
+    mHeartbeatTimeoutMs =
+        (int) Configuration.getMs(PropertyKey.MASTER_METRICS_HEARTBEAT_TIMEOUT_MS);
+    mLastSuccessfulHeartbeatMs = System.currentTimeMillis();
```

Comment:
Yes? This is instantiated at the client.

Reply:
We set the last successful heartbeat without actually doing a heartbeat. Since this is in the constructor, could there be an issue where `run` is not called immediately after the constructor causing the first heartbeat to incorrectly believe it has timed out?

Reply:
I see. Considering @aaudiber's point, this last successful heartbeat time is not that useful. I'll get rid of it.